csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1788,21 +1825,196 @@ class StackBasedSharedMemAllocator : kir::IrVisitor {
 
 } // namespace
 
+// Use allocation info map to find aliases, i.e. allocations that are properly
+// sized and parallelized so that they can be re-used without any
+// synchronization.
+std::vector<Expr*> aliasMemoryAllocations(
+    const std::vector<Expr*>& exprs,
+    AllocationInfoMap& allocation_info_map) {
+  ReusableAllocationFinder::find(exprs, allocation_info_map);
+  return AllocationAliasModifier::modify(exprs, allocation_info_map);
+}
+
+class PromoteReuseSyncModifier : private kir::ExprMutator {
+ public:
+  PromoteReuseSyncModifier(
+      const std::vector<Expr*>& exprs,
+      const AllocationInfoMap& allocation_info_map)
+      : allocation_info_map_(allocation_info_map) {
+    // Find next allocation after last aliased read of all allocations whose
+    // reuse we need to promote, and record shortest sync intervals relative to
+    // subsequent allocations.
+    for (const auto& alloc_info : allocation_info_map.allAllocationInfos()) {
+      auto tv = alloc_info->alloc_expr->buffer()->as<TensorView>();
+      if (tv->getMemoryType() != MemoryType::Shared ||
+          !tv->shouldPromoteReuse()) {
+        continue;
+      }
+      auto last_read = alloc_info->getAliasedOuterLastRead();
+
+      std::optional<int> nearest_first_write = std::nullopt;
+
+      for (const auto& other : allocation_info_map.allAllocationInfos()) {
```

Comment:
What happens if `other` is already aliased to `tv` or an other allocation? Shouldn't it be excluded from this analysis?